src/fs/metadata_log.cc

```diff
@@ -0,0 +1,326 @@
+        });
+    } else {
+        cut_out_data_range(file, new_data_vec.data_range);
+    }
+
+    file.data.emplace(new_data_vec.data_range.beg, std::move(new_data_vec));
+}
+
+void metadata_log::cut_out_data_range(inode_info::file& file, file_range range) {
+    file.cut_out_data_range(range, [](const inode_data_vec& data_vec) {
+        (void)data_vec; // TODO: for compaction: update used inode_data_vec
+    });
+}
+
+std::variant<inode_t, metadata_log::path_lookup_error> metadata_log::path_lookup(const sstring& path) const {
+    if (path.empty() or path[0] != '/') {
+        return path_lookup_error::NOT_ABSOLUTE;
+    }
+
+    std::vector<inode_t> components_stack = {_root_dir};
+    size_t beg = 0;
+    while (beg < path.size()) {
+        range component_range = {beg, path.find('/', beg)};
+        bool check_if_dir = false;
+        if (component_range.end == path.npos) {
+            component_range.end = path.size();
+            beg = path.size();
+        } else {
+            check_if_dir = true;
+            beg = component_range.end + 1; // Jump over '/'
+        }
+
+        // TODO: I don't like that we make a copy here -- it is totally redundant and inhibits adding noexcept
+        sstring component = path.substr(component_range.beg, component_range.size());
+        // Process the component
+        if (component == "") {
+            continue;
+        } else if (component == ".") {
+            assert(component_range.beg > 0 and path[component_range.beg - 1] == '/' and "Since path is absolute we do not have to check if the current component is a directory");
+            continue;
+        } else if (component == "..") {
+            if (components_stack.size() > 1) { // Root dir cannot be popped
+                components_stack.pop_back();
+            }
+        } else {
+            auto dir_it = _inodes.find(components_stack.back());
+            assert(dir_it != _inodes.end() and "inode comes from some previous lookup (or is a root directory) hence dir_it has to be valid");
+            assert(dir_it->second.is_directory() and "every previous component is a directory and it was checked when they were processed");
+            auto& curr_dir = dir_it->second.get_directory();
+
+            auto it = curr_dir.entries.find(component);
+            if (it == curr_dir.entries.end()) {
+                return path_lookup_error::NO_ENTRY;
+            }
+
+            inode_t entry_inode = it->second;
+            if (check_if_dir) {
+                auto entry_it = _inodes.find(entry_inode);
+                assert(entry_it != _inodes.end() and "dir entries have to exist");
+                if (not entry_it->second.is_directory()) {
+                    return path_lookup_error::NOT_DIR;
+                }
+            }
+
+            components_stack.emplace_back(entry_inode);
+        }
+    }
+
+    return components_stack.back();
+}
+
+future<inode_t> metadata_log::futurized_path_lookup(const sstring& path) const {
```

Comment:
This is the function we're more likely to use than raw `path_lookup`, right? I suggest naming this one `path_lookup`, and the other one `do_path_lookup`, it's a de-facto standard in seastar.